### ISSUE TITLE
Serialize AST according to CSSOM

### ISSUE THREAD

**Issue by Zirro** [csstree/csstree#52]
The [CSS Object Model](https://drafts.csswg.org/cssom/) specification includes many rules on how to serialize CSS in a consistent way. For example, the current translate() output `p{color:green}` would become `p { color: green; }` using the [CSSOM rules](https://drafts.csswg.org/cssom/#serialize-a-css-rule). It even covers parts like how to serialize a [string](https://drafts.csswg.org/cssom/#serialize-a-string).

I recently started working on an implementation of the CSSOM and would like to use CSSTree as the parser, thanks to the priority you put on parsing standard syntax. I'll probably write my own `generate` functions to follow these rules if you decline, but I figured I would ask if you want to make CSSTree serialize CSS according the CSSOM by default? Thanks!

**Comment by Zirro** [comment]
I have compared the output from CSSTree with [a web platform test](https://github.com/w3c/web-platform-tests/blob/master/cssom/serialize-values.html) for [serializing values](https://drafts.csswg.org/cssom/#serialize-a-css-component-value), and created a table based on the differences:

| Input                   | Output            | Expected               | Conclusion |
| ----------------|-----------------|----------------|-------------|
| `rgb(50, 75, 100)` | `rgb(50,75,100)` | `rgb(50, 75, 100)` | Commas should be followed by a space character
| `'Lucida Grande'` | `'Lucida Grande'` | `\"Lucida Grande\"` | Strings are serialized with escaped double quotes
| `url(http://localhost/)` | `url(http://localhost/)` | `url(\"http://localhost/\")` | URLs are serialized as strings and thus surrounded by escaped double quotes
| `5% .5%` `5% -.5%` | `5% .5%` `5% -.5%` | `5% 0.5%` `5% -0.5%` | Decimals begin with 0, rather than the dot
| `5% -0px` | `5% -0px` | `5% 0px` | 0 is not preceded by a minus sign
| `counter(par-num)` | `counter(par-num)` | `counter(par-num, decimal)` | Might be incorrect/outdated? Will see if I can confirm this later.

**Comment by lahmatiy** [comment]
Some notes:
- double quotes excaping needed when used inside a string only, otherwise it produces incorrect result
- not sure serialization should normalize numbers or url's or anything
- CSS generation from AST should not add anything additional that not in AST, so example with `counter` is incorrect; moreover we need to know more about values and functions for such transformation

In general, CSSOM serialization is not applicable for AST for each parts. Since this is for generation a string representation of an object, last one contains a normalized values not an original.

AST describes a source as close as posible (with some loses like an optional white spaces at the moment), and generator should generate a CSS by this AST (we manage the resulting CSS by form of AST). Main applications of AST are source-to-source transformations and analysis, so it should not to be normalized.

CSSTree was extracted from [CSSO](https://github.com/css/csso) (CSS optimizer), because it can be used for other packages and tasks. At the moment CSSO is main consumer of CSSTree, that's why generator produces a CSS in compact form. I working on generator improvements ([improve-generator](https://github.com/csstree/csstree/tree/improve-generator) branch) to make it easier to create any custom generator and define some modes (compact, pretty print etc).

What you describe is possible by AST transformation. So it's mostly a transformation (normalization) task rather than a generator.

**Comment by Zirro** [comment]
Indeed, I am asking for an option to normalize the output rather than the AST itself. I look forward to the improve-generator branch being merged, as that will surely make it easier.

In the meantime, I'll see if I can write my own CSSOM-based transformer for the CSSTree AST within my project. Feel free to close this issue if you'd prefer to not include a CSSOM-transformer as part of CSSTree itself.
